core/psh/psh.c

```diff
@@ -972,7 +974,117 @@ static void psh_signalstop(int sig)
 }
 
 
-static int psh_run(void)
+static int psh_login(void)
+{
+	int insz = -1, code;
+	const int maxinsz = 32;
+	char c, indata[maxinsz+1], indatasave[maxinsz+1], shdw[128];
+	struct termios mode;
+
+	tcgetattr(0, &mode);
+	mode.c_lflag |= (ICANON | ECHO);
+	tcsetattr(0, TCSANOW, &mode);
+
+	/* Get login from user */
+	memset(indata, '\0', sizeof(indata));
+
+	for (;;) {
+		if (insz < 0) {
+			fprintf(stdout, "Login: ");
+			fflush(stdout);
+			insz++;
+		}
+
+		if(read(STDIN_FILENO, &c, 1) != 1)
+			continue;
+			
+		if (c == '\n') {
+			if (insz <= maxinsz && insz > 0) {
+				memset(indatasave, '\0', sizeof(indatasave));
+				memcpy(indatasave, indata, sizeof(indatasave));
+				memset(indata, '\0', sizeof(indata));
+				break;
+			}
+			else {
+				memset(indata, '\0', sizeof(indata));
+				insz = -1;
+				continue;
+			}
+		}
+		else if (isprint(c)) {
+			if (insz < maxinsz)
+				indata[insz] = c;
+			insz++;
+			continue;
+		}
+	}
+
+	/* Get password from user */
+	memset(shdw, '\0', sizeof(shdw));
+	memset(indata, '\0', sizeof(indata));
+	mode.c_lflag &= ~(ECHO);
+	tcsetattr(0, TCSANOW, &mode);
+	insz = -1;
+	for (;;) {
+		if (insz < 0) {
+			printf("Password: ");
+			fflush(stdout);
+			insz++;
+		}
+
+		if(read(STDIN_FILENO, &c, 1) != 1)
+			continue;
+
+		if (c == '\n') {
+			memset(shdw, '\0', sizeof(shdw));
+			if (insz <= maxinsz) {
+				memcpy(shdw, crypt(indata, "ABCD"), sizeof(shdw)); 
```

Comment:
logic for password checking should not be part of input routine, move out of the `for` loop.

Also: invalid usage of `crypt` function. You really need to read other login implementations.
Also2: why do we need to copy the result of the `crypt`?

Reply:
After adding subroutine for credential getting the password checking is made after all data is acquired from user, not during input.

I checked it with busybox password handling and it is now made in similar fashion.